### PR TITLE
Add a Prisma adapter

### PR DESCRIPTION
**Pull request by timleslie**
[Prisma 2](https://www.prisma.io/blog/announcing-prisma-2-n0v98rzc8br1/) has recently been released. It provides a powerful javascript library for interfacing with databases. The Prisma team have recently received [significant funding](https://www.prisma.io/blog/prisma-raises-series-a-saks1zr7kip6/) to continue development of their library.

This PR introduces support for using Prisma as a database adapter within Keystone. 

We are still reviewing Prisma for feature completeness and correctness to ensure that all of Keystone's features will work as expected.

We are also investigating the Prisma migration framework in order to provide a better DX for developers who need to perform migrations on their database as their Keystone system evolves.

Note: This PR is still WIP, particularly in regard to how, where and when we are generating the Prisma client.

See also: https://github.com/prisma/prisma/issues/3265

#### Relationships

| Type | Create | Read | Count | Update | Delete | Notes |
| :---  | :---: | :---: | :---: | :---: | :---: | :--- |
| One-to-many one-sided | ✅  | ✅  | ✅  | ✅  | ✅  |
| One-to-many two-sided | ✅  | ✅  | ✅  | ✅  | ✅  | ~~https://github.com/prisma/prisma/issues/3141~~ |
| Many-to-many one-sided | ✅  | ✅  | ✅  | ✅  | ✅  |
| Many-to-many two-sided  | ✅  | ✅  | ✅  | ✅  | ✅  |
| One-to-one   | ✅  | ✅  | ☑️   | ✅  | ✅  |
| **Self-referential**
| One-to-many one-sided | ✅  | ✅  | ✅  | ✅  | ✅  |
| One-to-many two-sided | ✅ | ✅ | ✅  | ✅ | ✅ | ~~https://github.com/prisma/prisma/issues/3078~~ |
| Many-to-many one-sided | ✅  | ✅  | ✅  | ✅  | ✅  |
| Many-to-many two-sided | ✅  | ✅  | ✅  | ✅  | ✅  |
| One-to-one  | ✅ | ✅ | ☑️   | ✅  | ✅  | ~~https://github.com/prisma/prisma/issues/3153~~ |

#### Fields

| Name | Prisma Type | Core | Filter | Required | Unique | Notes |
| :--- | :---: | :---: | :---: | :---: | :---: | :--- |
| **Core**
| `CalendarDay` | `DateTime` | ✅ | ✅ | ✅  | ✅  |
| `Checkbox` | `Boolean` | ✅  | ✅  | ✅  | ☑️  |
| `DateTime` | `DateTime` + `String` | ✅  | ✅  | ✅ | ✅  |
| `DateTimeUtc` | `DateTime` | ✅  | ✅  | ✅  | ✅  |
| `Decimal` | `N/A`  | 🔴  | 🔴  | 🔴  | 🔴  | Prisma does not currently support a decimal database type (https://github.com/prisma/prisma/issues/3374) (https://github.com/prisma/prisma/issues/3447) |
| `File` | `Json` | ? | 🔴  | ✅  | ☑️  | https://github.com/prisma/prisma/issues/3579 |
| `Float` | `Float` | ✅   | ✅  | ✅ | ✅  |
| `Integer` | `Int` | ✅  | ✅  | ✅ | ✅  |
| `Password` | `String` | ? | ✅  | ✅  | ☑️  | Need to be able to check for a regex |
| `Relationship` | `@relation` | ✅  | ✅  | ☑️  | ☑️  |
| `Select` | `Enum` \| `Int` \| `String` | ✅  | ✅  | ✅  | ✅  |
| `Slug` | `String` | ✅ | ✅ | ✅ | ✅  |
| `Text` | `String` | ✅  | ✅   | ✅  | ✅  | ~~https://github.com/prisma/prisma-client-js/issues/690~~ |
| `Url` |`String` | ✅  | ✅   | ✅  | ✅  | ~~https://github.com/prisma/prisma-client-js/issues/690~~ |
| `Uuid` | `String` | ? | ✅  | ✅  | ✅  | We have used a `String` here for now, but we should really work out how to get an actual `uuid` column | 
| `Virtual` | ☑️  | ☑️  | ☑️  | ☑️ | ☑️  |
| **Extra**
| `Color` | `String` | ✅  | ✅  | ✅  | ✅ | ~~https://github.com/prisma/prisma-client-js/issues/690~~ |
| `Content` | `@relation` | ✅ | ✅ | ☑️  | ☑️  |
| `AuthedRelationship` | `@relation` | ✅ | ✅ | ☑️  | ☑️  |
| `AutoIncrement` | `Int` | ✅ | ✅ | ✅ | ✅ |
| `CloudinaryImage` | `Json` | ? | ? | ? | ☑️  |
| `GoogleLocation` | `Json` | ? | ? | ✅  | ☑️  |
| `Markdown` | `String` | ✅  | ✅  | ✅  | ✅  | ~~https://github.com/prisma/prisma-client-js/issues/690~~ |
| `MongoId` | `String` | ✅  | ✅  | ✅  | ✅  |
| `OEmbed` | `Json` | ? | ? | ? | ☑️  |
| `Unsplash` | `Json` | ? | ? | ? | ☑️  |
| `Wysiwyg` | `String` | ✅ | ✅ | ✅  | ✅  |

 * ✅  - Working
 * 🔴 - Not working (test failures)
 * ☑️ - N/A (working by default)
 * ❓ - Still being investigated

### CODE REVIEW

File: packages/fields-auto-increment/package.json [hunk 1]
```diff
@@ -12,6 +12,7 @@
   "dependencies": {
     "@babel/runtime": "^7.8.4",
     "@keystonejs/adapter-knex": "^11.0.0",
+    "@keystonejs/adapter-prisma": "^0.0.0",
```
Comment (gautamsi) [inline packages/fields-auto-increment/package.json:15]: This is what we should move away from. The way each adapter related code is required in each field (almost)

I wish we could have some spec with which the base adapter would expose the transformation functions and we use that in field and new adapter can provide such thing directly. 

if you look at base adapter in Keystone, that has bunch of high level api and each implementing adapter exposes them as private/internal functions.

Comment (JedWatson) [reply packages/fields-auto-increment/package.json:15]: I 100% agree with you here, but let's track / discuss that separately to this PR 👍

**Comment by VinayaSathyanarayana** [comment]
Prisma 2.5 has been released https://github.com/prisma/prisma/releases/tag/2.5.1

**Comment by timleslie** [comment]
> Prisma 2.5 has been released https://github.com/prisma/prisma/releases/tag/2.5.1

Yep, we're keeping up to date with all the releases 👍

File: packages/adapter-prisma/lib/adapter-prisma.js [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+            }
+          : value
+      ),
+      include: this._include(),
+    });
+  }
+
+  async _update(id, _data) {
+    const include = this._include();
+    const existingItem = await this.model.findOne({ where: { id: Number(id) }, include });
```
Comment (mitchellhamilton) [inline packages/adapter-prisma/lib/adapter-prisma.js:309]: This wouldn't work for string ids right?

(this comment is in reference to all the Number calls in this file, not just this one)

File: packages/adapter-prisma/lib/adapter-prisma.js [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+        // SELECT ... LIMIT <first>
+        ret.take = first;
+      }
+      if (skip !== undefined) {
+        // SELECT ... OFFSET <skip>
+        ret.skip = skip;
+      }
+      if (orderBy !== undefined) {
+        // SELECT ... ORDER BY <orderField>
+        const [orderField, orderDirection] = orderBy.split('_');
```
Comment (mitchellhamilton) [inline packages/adapter-prisma/lib/adapter-prisma.js:416]: This would break if you have a field with an _ in it and try to order by it right?

Comment (timleslie) [reply packages/adapter-prisma/lib/adapter-prisma.js:416]: Yep, we need to write some tests for this, because I'm pretty certain that this bug exists in the `knex` adapter as well.

File: packages/adapter-prisma/lib/adapter-prisma.js [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+        }
+
+        // FIXME: ask the field adapter if it supports the condition type
+        const supported =
+          fieldAdapter && fieldAdapter.getQueryConditions(fieldAdapter.dbPath)[condition];
+        if (supported) {
+          return supported(value);
+        } else {
+          // Many relationship
+          const [fieldPath, constraintType] = condition.split('_');
```
Comment (mitchellhamilton) [inline packages/adapter-prisma/lib/adapter-prisma.js:470]: field with `_` again right?

File: packages/adapter-prisma/lib/adapter-prisma.js (hunk 2 above)
Comment (timleslie) [reply packages/adapter-prisma/lib/adapter-prisma.js:309]: Yep, good catch, I need to loop back and remove the need for all these `Number` conversions, because they shouldn't be happening in this part of the adapter.

File: packages/fields/src/types/Password/Implementation.js [hunk 5]
```diff
@@ -170,3 +171,32 @@ export class KnexPasswordInterface extends CommonPasswordInterface(KnexFieldAdap
...
+
+  getPrismaSchema() {
+    return [this._schemaField({ type: 'String' })];
+  }
+
+  getQueryConditions(dbPath) {
+    // JM: I wonder if performing a regex match here leaks any timing info that
+    // could be used to extract information about the hash.. :/
+    return {
+      // FIXME: Prisma needs to support regex matching...
```
Comment (mitchellhamilton) [inline packages/fields/src/types/Password/Implementation.js:196]: Is there an issue about this? I tried searching on the prisma repo but couldn't find anything

Comment (timleslie) [reply packages/fields/src/types/Password/Implementation.js:196]: Not that I know of, it's something I need to bring up with them, I just need to take the time to formulate the question properly so I can explain exactly what the feature would look like.

File: packages/fields/src/types/Password/Implementation.js [hunk 6]
```diff
@@ -170,3 +171,32 @@ export class KnexPasswordInterface extends CommonPasswordInterface(KnexFieldAdap
...
 }
+
+export class PrismaPasswordInterface extends CommonPasswordInterface(PrismaFieldAdapter) {
+  constructor() {
+    super(...arguments);
+
+    // Error rather than ignoring invalid config
+    if (this.config.isUnique || this.config.isIndexed) {
+      throw (
+        `The Password field type doesn't support indexes on Knex. ` +
```
Comment (mitchellhamilton) [inline packages/fields/src/types/Password/Implementation.js:182]:
```suggestion
        `The Password field type doesn't support indexes on Prisma. ` +
```

File: examples/todo/index.js [hunk 7]
```diff
@@ -1,18 +1,27 @@
 const { Keystone } = require('@keystonejs/keystone');
-const { MongooseAdapter } = require('@keystonejs/adapter-mongoose');
+const { PrismaAdapter } = require('@keystonejs/adapter-prisma');
```
Comment (mitchellhamilton) [inline examples/todo/index.js:2]: We probs want to revert these changes before merging?

Comment (timleslie) [reply examples/todo/index.js:2]: Yep, this example will be completely reverted before the merge, this is just my scratch space right now 👍

File: packages/adapter-prisma/README.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+subSection: database-adapters
+title: Prisma adapter
+[meta]-->
+
+# Prisma database adapter
+
+[![View changelog](https://img.shields.io/badge/changelogs.xyz-Explore%20Changelog-brightgreen)](https://changelogs.xyz/@keystonejs/adapter-prisma)
+
+The [Prisma](https://www.prisma.io/) adapter is a general purpose adapter which can be used to connect to a range of different database backends.
+At present, the only fully tested backend is `Postgres`, however Knex gives the potential for `MSSQL`, `MySQL`, `MariaDB`, `SQLite3`, `Oracle`, and `Amazon Redshift` to be supported.
```
Comment (mitchellhamilton) [inline packages/adapter-prisma/README.md:12]:
```suggestion
At present, the only fully tested backend is `Postgres`, however Prisma gives the potential for `MSSQL`, `MySQL`, `MariaDB`, `SQLite3`, `Oracle`, and `Amazon Redshift` to be supported.
```

+I think those supported databases are wrong?

Comment (timleslie) [reply packages/adapter-prisma/README.md:12]: Almost the whole `README` is copy pasta from knex right now. Will fix that before we ship but probably not before we've tweaked the DX a little.

File: packages/fields-unsplash/src/Implementation.js [hunk 9]
```diff
@@ -294,3 +295,21 @@ export class KnexUnsplashInterface extends CommonUnsplashInterface(KnexFieldAdap
...
+
+export class PrismaUnsplashInterface extends CommonUnsplashInterface(PrismaFieldAdapter) {
+  constructor() {
+    super(...arguments);
+
+    // Error rather than ignoring invalid config
+    // We totally can index these values, it's just not trivial. See issue #1297
+    if (this.config.isIndexed) {
+      throw (
+        `The Unsplash field type doesn't support indexes on Knex. ` +
```
Comment (mitchellhamilton) [inline packages/fields-unsplash/src/Implementation.js:307]:
```suggestion
        `The Unsplash field type doesn't support indexes on Prisma. ` +
```

File: packages/fields-mongoid/src/index.js [hunk 10]
```diff
@@ -19,13 +19,23 @@ export const MongoId = {
...
+          `The MongoId field type doesn't provide a default primary key field configuration for knex. ` +
+          `You'll need to supply your own 'id' field for each list or use a different field type for your ` +
+          `ids (eg '@keystonejs/fields-auto-increment').`
+        );
+      },
+    },
+    prisma: {
+      getConfig: () => {
+        throw (
+          `The MongoId field type doesn't provide a default primary key field configuration for knex. ` +
```
Comment (mitchellhamilton) [inline packages/fields-mongoid/src/index.js:38]:
```suggestion
          `The MongoId field type doesn't provide a default primary key field configuration for Prisma. ` +
```

File: packages/fields-auto-increment/src/test-fixtures.js [hunk 11]
```diff
@@ -12,7 +12,7 @@ export const skipCreateTest = true;
 export const skipUpdateTest = true;
 
 // `AutoIncrement` field type is not supported by `mongoose`. So, we need to filter it out while performing `API` tests.
-export const unSupportedAdapterList = ['mongoose'];
+export const unSupportedAdapterList = ['mongoose', 'prisma'];
```
Comment (mitchellhamilton) [inline packages/fields-auto-increment/src/test-fixtures.js:15]: Why is Prisma here? Isn't AutoIncrement the default primary key for Prisma?

Comment (timleslie) [reply packages/fields-auto-increment/src/test-fixtures.js:15]: Oh, mostly I had this disabled because tests weren't passing for reasons I didn't have a chance to look into. Let me put this to the top of my fields matrix work and I should be able to get this field type with all green ticks now 👍

File: packages/adapter-prisma/lib/adapter-prisma.js [hunk 12]
```diff
@@ -0,0 +1,611 @@
...
+  async dropDatabase() {
+    let migrationNeeded = true;
+    for (const { tablename } of await this.prisma.$queryRaw(
+      `SELECT tablename FROM pg_tables WHERE schemaname='${this.dbSchemaName}'`
+    )) {
+      if (tablename.includes('_Migration')) {
+        migrationNeeded = false;
+      }
+      await this.prisma.$queryRaw(
+        `TRUNCATE TABLE \"${this.dbSchemaName}\".\"${tablename}\" CASCADe;`
```
Comment (mitchellhamilton) [inline packages/adapter-prisma/lib/adapter-prisma.js:212]: Is `CASCADe` meant to be `CASCADE`?

Comment (timleslie) [reply packages/adapter-prisma/lib/adapter-prisma.js:212]: Yep :-)

File: packages/fields-auto-increment/src/test-fixtures.js (hunk 11 above)
Comment (timleslie) [reply packages/fields-auto-increment/src/test-fixtures.js:15]: Ah yes, now I recall, these failures are related to the `Number` stuff throughout the adapter code. Need to fix both of those together.

File: packages/adapter-prisma/lib/adapter-prisma.js [hunk 13]
```diff
@@ -0,0 +1,611 @@
...
+
+  _runPrismaCmd(cmd) {
+    return execSync(`yarn prisma ${cmd} --schema ${this.schemaPath}`, {
+      env: { ...process.env, DATABASE_URL: this._url() },
+      encoding: 'utf-8',
+    });
+  }
+
+  disconnect() {
+    this.prisma.$disconnect();
```
Comment (mitchellhamilton) [inline packages/adapter-prisma/lib/adapter-prisma.js:237]: I assume this returns a promise so we should return it?

Comment (timleslie) [reply packages/adapter-prisma/lib/adapter-prisma.js:237]: Yep, good call 👍

File: packages/adapter-prisma/lib/adapter-prisma.js [hunk 14]
```diff
@@ -0,0 +1,611 @@
...
+    this.url = this.config.url || process.env.DATABASE_URL;
+  }
+
+  async _connect({ rels }) {
+    await this._generateClient(rels);
+    const { PrismaClient } = require(this.clientPath);
+    this.prisma = new PrismaClient({
+      log: this.enableLogging && ['query'],
+      datasources: { postgresql: { url: this._url() } },
+    });
```
Comment (mitchellhamilton) [inline packages/adapter-prisma/lib/adapter-prisma.js:26]: Shouldn't we explicitly call and await .$connect() here so we're ready right away rather than on the first call to prisma?

Comment (timleslie) [reply packages/adapter-prisma/lib/adapter-prisma.js:26]: Yep, we can do that 👍

File: packages/adapter-prisma/lib/adapter-prisma.js (hunk 2 above)
Comment (timleslie) [reply packages/adapter-prisma/lib/adapter-prisma.js:317]: For now I'm going to say that we only support `AutoIncrement` as the field type for the `id` field. We need to introduce better testing for the other cases before we decide how best to clean this stuff up.

File: packages/adapter-prisma/lib/adapter-prisma.js (hunk 3 above)
Comment (timleslie) [reply packages/adapter-prisma/lib/adapter-prisma.js:424]: I'm going to leave this as an open issue until someone can write some tests and we fix this across all adapters.

File: packages/adapter-prisma/lib/adapter-prisma.js (hunk 4 above)
Comment (timleslie) [reply packages/adapter-prisma/lib/adapter-prisma.js:478]: Will fix when tests exist across all adapters 👍

File: packages/fields-auto-increment/src/test-fixtures.js (hunk 11 above)
Comment (timleslie) [reply packages/fields-auto-increment/src/test-fixtures.js]: Fixed.

File: packages/fields-cloudinary-image/src/Implementation.js [hunk 15]
```diff
@@ -93,5 +93,11 @@ class CloudinaryImage extends File.implementation {
 
 const MongoCloudinaryImageInterface = File.adapters.mongoose;
 const KnexCloudinaryImageInterface = File.adapters.knex;
+const PrismaCloudinaryImageInterface = File.adapters.knex;
```
Comment (mitchellhamilton) [inline packages/fields-cloudinary-image/src/Implementation.js:96]:
```suggestion
const PrismaCloudinaryImageInterface = File.adapters.prisma;
```

Comment (timleslie) [reply packages/fields-cloudinary-image/src/Implementation.js:96]: Nice catch 👍

**Review by mitchellhamilton — APPROVED** [review]
